cmd/tlsproxy/tlsproxy.go

```diff
@@ -0,0 +1,214 @@
+	}
+	if *servingkey != "" && *servingcert != "" {
+		if _, err := os.Stat(*servingkey); os.IsNotExist(err) {
+			return append(errs, errors.New(fmt.Sprintf("file %s does not exist", *servingkey)))
+		}
+		if _, err := os.Stat(*servingcert); os.IsNotExist(err) {
+			return append(errs, errors.New(fmt.Sprintf("file %s does not exist", *servingcert)))
+		}
+	}
+
+	return errs
+}
+
+func (c *config) Init() error {
+	logrus.SetLevel(log.SanitizeLogLevel(*logLevel))
+	logrus.SetFormatter(&logrus.TextFormatter{FullTimestamp: true})
+	c.log = logrus.NewEntry(logrus.StandardLogger())
+
+	c.username = os.Getenv("USERNAME")
+	c.password = os.Getenv("PASSWORD")
+
+	// validate flags
+	if errs := c.validate(); len(errs) > 0 {
+		return errors.Wrap(kerrors.NewAggregate(errs), "cannot validate flags")
+	}
+
+	// sanitize inputs
+	var err error
+	c.redirectURL, err = url.Parse(*hostname)
+	if err != nil {
+		return err
+	}
+
+	c.whitelistRegexp, err = regexp.Compile(*whitelist)
+	if err != nil {
+		return err
+	}
+
+	b, err := ioutil.ReadFile(*cacert)
+	if err != nil {
+		return err
+	}
+	pool := x509.NewCertPool()
+	pool.AppendCertsFromPEM(b)
+	cert, err := tls.LoadX509KeyPair(*cert, *key)
+	if err != nil {
+		return err
+	}
+
+	c.cli = &http.Client{
+		Transport: &http.Transport{
+			TLSClientConfig: &tls.Config{
+				InsecureSkipVerify: *insecure,
+				RootCAs:            pool,
+				Certificates:       []tls.Certificate{cert},
+			},
+		},
+	}
+
+	return nil
+}
+
+func usage() {
+	fmt.Printf("Usage:\n")
+	fmt.Printf("\"%s -hostname\" url to rewrite \n\n", os.Args[0])
+	flag.PrintDefaults()
+	os.Exit(1)
+}
+
+func (c *config) Run() error {
+	handlerFunc := http.HandlerFunc(func(rw http.ResponseWriter, req *http.Request) {
+		req.URL.Scheme = c.redirectURL.Scheme
+		req.URL.Host = c.redirectURL.Host
+		req.RequestURI = ""
+		req.Host = ""
+
+		if !c.whitelistRegexp.MatchString(req.URL.String()) || req.Method != http.MethodGet {
+			http.Error(rw, http.StatusText(http.StatusForbidden), http.StatusForbidden)
+			return
+		}
+
+		// check authentication
+		if c.username != "" {
+			if !c.checkAuth(rw, req) {
+				http.Error(rw, http.StatusText(http.StatusForbidden), http.StatusForbidden)
+				return
+			}
+		}
+
+		resp, err := c.cli.Do(req)
+		if err != nil {
+			http.Error(rw, err.Error(), http.StatusInternalServerError)
+			return
+		}
+		defer resp.Body.Close()
+
+		for k, v := range resp.Header {
+			rw.Header()[k] = v
+		}
+		rw.WriteHeader(resp.StatusCode)
+		io.Copy(rw, resp.Body)
+	})
+
+	if *servingcert != "" && *servingkey != "" {
+		c.log.Debug("starting in reencrypt mode")
+		return http.ListenAndServeTLS(*listen, *servingcert, *servingkey, handlerFunc)
+	}
+	c.log.Debug("starting in plain text mode")
+	return http.ListenAndServe(*listen, handlerFunc)
+}
+
+func (c *config) checkAuth(rw http.ResponseWriter, req *http.Request) bool {
+	if username, password, ok := req.BasicAuth(); ok {
+		return username == c.username && password == c.password
```

Comment:
```
username, password, _ := req.BasicAuth()
return username == c.username && password == c.password
```

is clearer.  Also name the function something like authIsOK so that usage is clearer at the caller